Python/pytime.c

```diff
@@ -35,6 +35,59 @@ error_time_t_overflow(void)
                     "timestamp out of range for platform time_t");
 }
 
+static void
+_PyTime_overflow(void)
+{
+    PyErr_SetString(PyExc_OverflowError,
+                    "timestamp too large to convert to C _PyTime_t");
+}
+
+#if defined(__APPLE__) || defined(MS_WINDOWS)
+/* Euclidean algorithm to compute the greatest common divisor of a and b. */
+static _PyTime_t
+_PyTime_GCD(_PyTime_t a, _PyTime_t b)
+{
+    _PyTime_t x, y;
+
+    if (a < b) {
+        x = b;
+        y = a;
+    }
+    else {
+        x = a;
+        y = b;
+    }
+    assert(x >= 1);
+    assert(y >= 1);
+    while (y != 0) {
+        _PyTime_t y2 = y;
+        y = x % y;
+        x = y2;
+    }
+    return x;
+}
+
+/* Compute tp * mul / div.
+   Raise an exception and return -1 on integer overflow.
+   Return 0 on success. */
+static int
+_PyTime_MulDiv(_PyTime_t *tp, _PyTime_t mul, _PyTime_t div)
+{
+    assert(mul >= 1);
+    assert(div >= 1);
+
+    _PyTime_t t = *tp;
+    if (t > _PyTime_MAX / mul) {
```

Comment:
It occurs to me that if the low order bits in `mul` are zero, and the high order bits in `div` are also zero, then shifting `mul` right and `div` left could allow even more computations to avoid overflow.

I'd be surprised if that was worth the extra complexity given the GCD calculations further down, though.

Reply:
Using GCD, there is no need to count the number of zero low order bits.

> I'd be surprised if that was worth the extra complexity given the GCD calculations further down, though.

On Windows, the ratio is SEC_TO_NS / perf_frequency where SEC_TO_NS=1_000_000_000. On my Windows 8.1 VM, perf_frequency=10_000_000. At the end, you get numer=100 and denom=1.

If you only use binary shifts, you loose the common divisor 5.